// @pythnetwork/pyth-evm-js/dist/cjs/tracer/debug-trace-call.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get debugTraceCallAction () {
        return debugTraceCallAction;
    },
    get extractPythPriceFeedsFromDebugTraceCall () {
        return extractPythPriceFeedsFromDebugTraceCall;
    }
});
const _viem = require("viem");
const _pythabi = require("../pyth-abi.cjs");
function extractPythPriceFeedsFromDebugTraceCall(trace, pythContractAddress, ignoreParsingErrors = false) {
    const result = new Set();
    if ((0, _viem.isAddressEqual)(trace.to, pythContractAddress)) {
        // Decode the calldata to see what function is being called
        try {
            const decoded = (0, _viem.decodeFunctionData)({
                abi: _pythabi.IPythAbi,
                data: trace.input
            });
            let priceFeedId;
            switch(decoded.functionName){
                case "getPrice":
                case "getPriceNoOlderThan":
                case "getPriceUnsafe":
                case "getEmaPrice":
                case "getEmaPriceNoOlderThan":
                case "getEmaPriceUnsafe":
                    {
                        priceFeedId = decoded.args[0];
                        break;
                    }
                default:
                    {
                        break;
                    }
            }
            if (priceFeedId !== undefined) {
                result.add(priceFeedId);
            }
        } catch (error) {
            if (!ignoreParsingErrors) {
                const thrownError = new Error(`Failed to decode calldata: ${trace.input}. Make sure correct Pyth contract address is used.`);
                thrownError.cause = error;
                throw thrownError;
            }
        }
    }
    if (trace.calls === undefined) {
        return result;
    }
    return new Set([
        ...result,
        ...trace.calls.flatMap((call)=>[
                ...extractPythPriceFeedsFromDebugTraceCall(call, pythContractAddress)
            ])
    ]);
}
const debugTraceCallAction = (client)=>({
        async debugTraceCall (args) {
            return client.request({
                method: "debug_traceCall",
                params: [
                    (0, _viem.formatTransactionRequest)(args),
                    "latest",
                    {
                        tracer: "callTracer"
                    }
                ]
            });
        }
    });
