// @pythnetwork/pyth-evm-js/dist/esm/tracer/debug-trace-call.mjs
import { decodeFunctionData, formatTransactionRequest, isAddressEqual } from "viem";
import { IPythAbi } from "../pyth-abi.mjs";
/**
 * Extract Pyth price feed IDs from a transaction call trace.
 */ export function extractPythPriceFeedsFromDebugTraceCall(trace, pythContractAddress, ignoreParsingErrors = false) {
    const result = new Set();
    if (isAddressEqual(trace.to, pythContractAddress)) {
        // Decode the calldata to see what function is being called
        try {
            const decoded = decodeFunctionData({
                abi: IPythAbi,
                data: trace.input
            });
            let priceFeedId;
            switch(decoded.functionName){
                case "getPrice":
                case "getPriceNoOlderThan":
                case "getPriceUnsafe":
                case "getEmaPrice":
                case "getEmaPriceNoOlderThan":
                case "getEmaPriceUnsafe":
                    {
                        priceFeedId = decoded.args[0];
                        break;
                    }
                default:
                    {
                        break;
                    }
            }
            if (priceFeedId !== undefined) {
                result.add(priceFeedId);
            }
        } catch (error) {
            if (!ignoreParsingErrors) {
                const thrownError = new Error(`Failed to decode calldata: ${trace.input}. Make sure correct Pyth contract address is used.`);
                thrownError.cause = error;
                throw thrownError;
            }
        }
    }
    if (trace.calls === undefined) {
        return result;
    }
    return new Set([
        ...result,
        ...trace.calls.flatMap((call)=>[
                ...extractPythPriceFeedsFromDebugTraceCall(call, pythContractAddress)
            ])
    ]);
}
export const debugTraceCallAction = (client)=>({
        async debugTraceCall (args) {
            return client.request({
                method: "debug_traceCall",
                params: [
                    formatTransactionRequest(args),
                    "latest",
                    {
                        tracer: "callTracer"
                    }
                ]
            });
        }
    });
