// @pythnetwork/pyth-evm-js/dist/esm/tracer/trace-call-many.mjs
import { decodeFunctionData, formatTransactionRequest, isAddressEqual } from "viem";
import { IPythAbi } from "../pyth-abi.mjs";
/**
 * Extract Pyth price feed IDs from a transaction call trace.
 */ export function extractPythPriceFeedsFromTraceCallMany(traceResults, pythContractAddress, ignoreParsingErrors = false) {
    const result = new Set();
    for (const traceResult of traceResults){
        for (const trace of traceResult.trace){
            if (isAddressEqual(trace.action.to, pythContractAddress)) {
                // Decode the calldata to see what function is being called
                try {
                    const decoded = decodeFunctionData({
                        abi: IPythAbi,
                        data: trace.action.input
                    });
                    let priceFeedId;
                    switch(decoded.functionName){
                        case "getPrice":
                        case "getPriceNoOlderThan":
                        case "getPriceUnsafe":
                        case "getEmaPrice":
                        case "getEmaPriceNoOlderThan":
                        case "getEmaPriceUnsafe":
                            {
                                priceFeedId = decoded.args[0];
                                break;
                            }
                        default:
                            {
                                break;
                            }
                    }
                    if (priceFeedId !== undefined) {
                        result.add(priceFeedId);
                    }
                } catch  {
                    if (!ignoreParsingErrors) {
                        throw new Error(`Failed to decode calldata: ${trace.action.input}. Make sure correct Pyth contract address is used.`);
                    }
                }
            }
        }
    }
    return result;
}
export const traceCallManyAction = (client)=>({
        async traceCallMany (args) {
            return client.request({
                method: "trace_callMany",
                params: [
                    args.map((a)=>[
                            formatTransactionRequest(a),
                            [
                                "trace"
                            ]
                        ]),
                    "latest"
                ]
            });
        }
    });
