// @pythnetwork/pyth-evm-js/dist/cjs/multicall3-bundler.cjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get MULTICALL3_ABI () {
        return MULTICALL3_ABI;
    },
    get MULTICALL3_ADDRESS () {
        return MULTICALL3_ADDRESS;
    },
    get multicall3Bundler () {
        return multicall3Bundler;
    }
});
const _viem = require("viem");
const MULTICALL3_ADDRESS = "0xcA11bde05977b3631167028862bE2a173976CA11";
const MULTICALL3_ABI = [
    {
        inputs: [
            {
                components: [
                    {
                        name: "target",
                        type: "address"
                    },
                    {
                        name: "allowFailure",
                        type: "bool"
                    },
                    {
                        name: "value",
                        type: "uint256"
                    },
                    {
                        name: "callData",
                        type: "bytes"
                    }
                ],
                name: "calls",
                type: "tuple[]"
            }
        ],
        name: "aggregate3Value",
        outputs: [
            {
                components: [
                    {
                        name: "success",
                        type: "bool"
                    },
                    {
                        name: "returnData",
                        type: "bytes"
                    }
                ],
                name: "returnData",
                type: "tuple[]"
            }
        ],
        stateMutability: "payable",
        type: "function"
    }
];
function multicall3Bundler(pythUpdate, call) {
    // Encode the multicall3 aggregate3 function call
    const multicallData = (0, _viem.encodeFunctionData)({
        abi: MULTICALL3_ABI,
        functionName: "aggregate3Value",
        args: [
            [
                {
                    target: pythUpdate.call.to,
                    allowFailure: false,
                    value: pythUpdate.call.value ?? 0n,
                    callData: pythUpdate.call.data ?? "0x"
                },
                {
                    target: call.to,
                    allowFailure: false,
                    value: call.value ?? 0n,
                    callData: call.data ?? "0x"
                }
            ]
        ]
    });
    // Calculate total value needed
    const totalValue = (call.value ?? 0n) + (pythUpdate.call.value ?? 0n);
    // Create the bundled transaction that calls multicall3
    return {
        to: MULTICALL3_ADDRESS,
        data: multicallData,
        value: totalValue,
        from: call.from
    };
}
